lang/java/reef-utils/src/test/java/org/apache/reef/util/MultiAsyncToSyncTest.java

```diff
@@ -0,0 +1,240 @@
+    this.sleepTimeMillis = sleepTimeMillis;
+    this.input = input;
+    this.identifier = identifier;
+    this.blocker = blocker;
+  }
+
+  /**
+   * Sleep and then ncrement the input value by one.
+   * @return The input value of the operation incremented by one.
+   * @throws Exception
+   */
+  public Integer call() throws Exception {
+    LOG.log(Level.INFO, "Sleeping...");
+    Thread.sleep(sleepTimeMillis);
+    LOG.log(Level.INFO, "Releasing caller...");
+    blocker.release(identifier);
+    return input + 1;
+  }
+}
+
+/**
+ * Use the MultiAyncToSync class to implement a synchronous API
+ * that uses asynchronous processing internally.
+ */
+final class SynchronousApi {
+  private static final Logger LOG = Logger.getLogger(SynchronousApi.class.getName());
+  private final int incrementerSleepTimeMillis;
+  private final MultiAsyncToSync blocker;
+  private final ExecutorService executor;
+  private final ConcurrentLinkedQueue<FutureTask<Integer>> taskQueue = new ConcurrentLinkedQueue<>();
+  private AtomicLong idCounter = new AtomicLong(0);
+
+  /**
+   * Parameterize the object as to length of processing time and call timeout.
+   * @param incrementerSleepTimeSeconds Length of time the incrementer sleeps before
+   *                                    performing the increment and returning.
+   * @param timeoutPeriodSeconds The length of time before the call will timeout.
+   */
+  SynchronousApi(final int incrementerSleepTimeSeconds, final long timeoutPeriodSeconds) {
+    this.incrementerSleepTimeMillis = 1000 * incrementerSleepTimeSeconds;
+    this.blocker = new MultiAsyncToSync(timeoutPeriodSeconds, SECONDS);
+    this.executor = Executors.newFixedThreadPool(2);
+  }
+
+  /**
+   * The one API call in our interface.
+   * @param input
+   * @return The input parameter incremented by one or zero for a timeout.
+   */
+  public Integer apiCall(final Integer input) {
+    // Create a future to run the asynchronous processing.
+    final long identifier = idCounter.getAndIncrement();
+    final FutureTask<Integer> task =
+        new FutureTask(new AsynchronousIncrementer(input, identifier, incrementerSleepTimeMillis, blocker));
+    taskQueue.add(task);
+
+    LOG.log(Level.INFO, "Running the incrementer...");
+    // Start the task and block the caller until it completes.
+    executor.execute(task);
+    if (blocker.block(identifier)) {
+      LOG.log(Level.INFO, "Call timed out...");
+      // Timeout occurred before the asynchronous processing completed.
+      return 0;
+    }
+    Integer result = 0;
+    try {
+      LOG.log(Level.INFO, "Call getting task result...");
+      result = task.get();
+    } catch (Exception e) {
+      result = -1;
+    }
+    return result;
+  }
+
+  /**
+   * Insure all test tasks have completed.
+   */
+  public void complete() {
+    try {
+      for (FutureTask<Integer> task : taskQueue) {
+        task.get();
+      }
+    } catch (Exception e) {
```

Comment:
I ended up creating an InvalidBlockedCallerIdentiferException and checking for it.